Allow for multiple simultaneous xfem cuts in an element

austinltam opened the pull request:
ref #15115

Changed the way split works on 2D elements. Split now uses cut plane ids associated to nodes at the element level. Node cut planes ids are assigned based on calls to addEdgeCut, addNodeCut, and addFragmentCut. Also allows for intersections of cutting planes inside elements, creating faceNodes at these intersections.

austinltam commented:
@bwspenc Here is my pull request. Thank you for the help!

bwspenc reviewed (commented):
That non-unity build failure is because you don't have the `#include`s quite right. You can sort it out locally by doing `MOOSE_UNITY=false make`.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/660476/) on 153ab5a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16705/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/664059/) on 92005a0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f0a1f1cbbec9e0ae7d6989835de9f7717b054f1e`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/669024/) on f9c8420 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9bd8b7ef542c75e40e816c9b89f33f8984900f2d`

austinltam commented:
@bwspenc I am not sure why the formatting is failing but it doesn't say it makes any changes when I run format-clang.

bwspenc commented:
@austinltam, since this has been sitting for a while, can you please re-base and push an update to this?

bwspenc commented on modules/xfem/test/tests/solid_mechanics_basic/tests:
```diff
@@ -111,6 +111,15 @@
+    map = false
+    # XFEM requires --enable-unique-ids in libmesh
+    unique_id = true
+    requirement = 'The XFEM system shall permit branched cracks to be represented in 2D by instantly cutting a 4-noded quadrilateral element by two prescribed evolving cutting planes'
```
instantly -> simultaneously

bwspenc commented on framework/src/meshgenerators/FancyExtruderGenerator.C:
```diff
@@ -562,7 +562,7 @@ FancyExtruderGenerator::generate()
   }
 
-  mesh->set_isnt_prepared();
+  mesh->is_prepared();
```
Did you mean to make this change? Seems unrelated.

bwspenc commented on modules/xfem/include/efa/EFAElement2D.h:
```diff
@@ -33,6 +33,10 @@ class EFAElement2D : public EFAElement
   std::vector<std::vector<EFAElement2D *>> _edge_neighbors;
   std::vector<EFAFragment2D *> _fragments;
   std::vector<EFAPoint> _local_node_coor;
+  unsigned int _current_cut_plane_idx = 0;
```
I know a lot of our code isn't documented like this, but as we add stuff, our convention is to add short doxygen comments for each variable in the header file. I realize this is kind of stating the obvious in this case, but something like this:
```
/// Index of the current cut plane
unsigned int _current_cut_plane_idx = 0;

bwspenc replied:
Also, delete the commented out stuff that you didn't end up using both here and below.

bwspenc commented on modules/xfem/include/efa/EFAElement2D.h:
```diff
@@ -127,6 +131,26 @@ class EFAElement2D : public EFAElement
 
   std::vector<EFANode *> getCommonNodes(const EFAElement2D * other_elem) const;
 
+  void addInteriorNode(EFAFaceNode * faceNode);
```
Doxygen comments here too

bwspenc commented on modules/xfem/include/efa/EFAElement2D.h:
```diff
@@ -127,6 +131,26 @@ class EFAElement2D : public EFAElement
   std::vector<EFANode *> getCommonNodes(const EFAElement2D * other_elem) const;
 
+  void addInteriorNode(EFAFaceNode * faceNode);
+  bool isInteriorNode(EFANode * node) const;
```
Functions like this should be documented using the function-style doxygen comments with brief descriptions of what the functions do and what the arguments to them are. Describe the parameters with `@param` and the return (if applicable) with `@return`. Look at the declaration of `addGeomMarkedElem2D` in XFEM.h for an example.

bwspenc commented on modules/xfem/include/efa/EFAElement3D.h:
```diff
@@ -155,6 +155,7 @@ class EFAElement3D : public EFAElement
                           bool add_to_neighbor,
                           bool add_to_adjacent);
   std::vector<EFANode *> getCommonNodes(const EFAElement3D * other_elem) const;
+  unsigned int getNewCutPlaneIdx();
```
Document this one

bwspenc commented on modules/xfem/include/efa/EFAFragment2D.h:
```diff
@@ -30,6 +30,8 @@ class EFAFragment2D : public EFAFragment
 private:
   EFAElement2D * _host_elem;
   std::vector<EFAEdge *> _boundary_edges;
+  std::vector<unsigned int> _frag_cut_plane_idx;
```
Add documentation

bwspenc commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -581,6 +582,7 @@ XFEM::markCutEdgesByState(Real time)
        ++pmeit)
   {
     const Elem * elem = pmeit->first;
+
```
Remove extra blank line.

bwspenc commented on modules/xfem/src/base/XFEM.C:
```diff
@@ -420,11 +419,13 @@ XFEM::markCutEdgesByGeometry()
-            if (!isElemAtCrackTip(gme.first))
-              _has_secondary_cut = true;
+            //            if (!isElemAtCrackTip(gme.first))
+            //              _has_secondary_cut = true;
```
If this is a change we really want to make, just delete the lines.

bwspenc commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -406,3 +406,31 @@ EFAEdge::removeEmbeddedNode(EFANode * node)
+    hasNode = true;
+
+  //  //check embedded nodes for node
+  //  for (unsigned int inode = 0; inode < _embedded_nodes.size(); ++inode)
```
Are there no embedded nodes on the edge? If you don't need this code, delete it.

bwspenc commented on modules/xfem/src/efa/EFAElement2D.C:
```diff
@@ -466,6 +468,7 @@ EFAElement2D::clearNeighbors()
 void
 EFAElement2D::setupNeighbors(std::map<EFANode *, std::set<EFAElement *>> & InverseConnectivityMap)
 {
+
```
Remove extra new line

bwspenc commented on modules/xfem/src/efa/EFAElement2D.C:
```diff
@@ -593,6 +596,7 @@ EFAElement2D::initCrackTip(std::set<EFAElement *> & CrackTipElements)
 unsigned int
 EFAElement2D::getCrackTipSplitElementID() const
 {
+
```
remove line

bwspenc commented on modules/xfem/src/efa/EFAElement2D.C:
```diff
@@ -836,8 +840,8 @@ EFAElement2D::updateFragments(const std::set<EFAElement *> & CrackTipElements,
-  if (num_cut_frag_edges > 3)
-    EFAError("In element ", _id, " there are more than 2 cut fragment edges");
+  //  if (num_cut_frag_edges > 3)
+  //    EFAError("In element ", _id, " there are more than 2 cut fragment edges");
```
Remove

bwspenc commented on modules/xfem/src/efa/EFAElement2D.C:
```diff
@@ -2067,3 +2099,331 @@ EFAElement2D::getCommonNodes(const EFAElement2D * other_elem) const
+  for (unsigned int i = 0; i < _interior_nodes.size(); ++i)
+  {
+    if (node == _interior_nodes[i]->getNode())
+    {
```
Our convention is to not have curly braces around one-liner blocks.

bwspenc commented on modules/xfem/src/efa/EFAElement2D.C:
```diff
@@ -2067,3 +2099,331 @@ EFAElement2D::getCommonNodes(const EFAElement2D * other_elem) const
+  {
+    interiorNode = _interior_nodes[i]->getNode();
+    if (node == interiorNode)
+    {
```
Remove braces around one liner.

bwspenc commented on modules/xfem/src/efa/EFAElement2D.C:
```diff
@@ -2067,3 +2099,331 @@ EFAElement2D::getCommonNodes(const EFAElement2D * other_elem) const
+  }
+}
+
+// void
```
Remove commented code

bwspenc reviewed (commented):
These comments are on mostly superficial formatting things. I'll need to pull this down and dig deeper into this, but go ahead and start doing these clean ups across your code.

austinltam committed:
adding tracking for multiple cuts per element and state based branching ref #15115

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/914405/) on 2c99fa4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7794ad7aaf012a2e54bdaf98ea3ebc722a794f3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/914516/) on 648a8bf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7794ad7aaf012a2e54bdaf98ea3ebc722a794f3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/914679/) on d5f5588 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7794ad7aaf012a2e54bdaf98ea3ebc722a794f3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/914890/) on b05b4ee wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7794ad7aaf012a2e54bdaf98ea3ebc722a794f3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/914959/) on 9b12de3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7794ad7aaf012a2e54bdaf98ea3ebc722a794f3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/937622/) on 61b454b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6b91c65f13fd6d7f53eab8838f08dd502df9e235`

austinltam committed:
Updated Crack Branching method and added parameters to turn it on and off

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/951852/) on aef2c58 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format db314bead5d74ab94880069dfbcd0967183f2805`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/951994/) on aef2c58 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format db314bead5d74ab94880069dfbcd0967183f2805`

bwspenc commented:
Rebased and force-pushed to trigger a new build against the updated libmesh that was just put in. Hopefully this will fix that build failure on the set_union.

austinltam committed:
code documentation and clean up per comments, also changed tolerance on new test
Small tweaks to update code to current revision and formatting fixes
code documentation and clean up per comments, also changed tolerance on new test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1011020/) on 0d22a0a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3587c9b242493465a46a69beb02b09ef93da0b01`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1011079/) on 319b3f9 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>3587c9</th>
      <th colspan=4>#16705 319b3f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.87%</td>
      <td>81.53%</td>
      <td>-0.34%</td>
      <td>69.51%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6344</td>
      <td>6612</td>
      <td>+268</td>
      <td>358</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1405</td>
      <td>1498</td>
      <td>+93</td>
      <td>157</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/16705/coverage_diff/xfem.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/16705/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/16705/coverage/xfem)

</details>

## Warnings

- `xfem` new line coverage rate 69.51% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Mac stack](https://civet.inl.gov/job/1011053/) on 319b3f9 : invalidated by @bwspenc

moosebuild commented:
Job [End user mac](https://civet.inl.gov/job/1011074/) on 319b3f9 : invalidated by @bwspenc

bwspenc commented:
Commenting to keep this open. @austinltam I know you have some updates for this. Just a reminder to push those when you get a chance.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1163135/) on 4e13701 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16705/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16705/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fa664410cea12f16d74333e1a608071309ee9ef2`

bwspenc commented:
@austinltam I ran clang-format on your commit, fixed XFEM.C to have unix rather than dos end lines, rebased it against the latest devel branch, and pushed that update just now.

austinltam committed:
:updated EFAfragment2D to use simple split in single cut cases and fixed node test error

bwspenc commented on modules/xfem/include/efa/EFAElement2D.h:
```diff
@@ -127,6 +133,96 @@ class EFAElement2D : public EFAElement
+  void reassignNodeToCutPlaneIdx(EFANode * node, unsigned int cut);
+  /**
+   * Get all cut plane nodes in a vector of vector
+   * @return vector of vectors containing all cut nodes binned by cut plane index
```
'binned'->'ordered'

bwspenc commented on modules/xfem/include/efa/EFAFragment2D.h:
```diff
@@ -51,5 +55,59 @@ class EFAFragment2D : public EFAFragment
+   * @param EmbeddedNodes is used to create new intersection nodes in the case of cut plane
+   * intersections
+   * @param vector of fragment cut plane ids
+   * @param vector of vectors with each cut node binned by cut planes
```
'binned'->'ordered'

bwspenc commented on modules/xfem/include/efa/EFAFragment2D.h:
```diff
@@ -51,5 +55,59 @@ class EFAFragment2D : public EFAFragment
+   * @param two length vector with coordinates of intersection point if it exists
+   * @return bool indicating if lines intersect
+   */
+  bool IntersectSegmentWithCutLine(const std::vector<double> & segment_point1,
```
Start method name with lower case to follow MOOSE style. (`intersectSegmentWithCutLine`)

bwspenc replied:
A few of these methods already exist in XFEMFuncs.C. Unless there's a really good reason to reimplement them here, you should use those, or add missing functionality to them if needed.
